cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -695,6 +702,75 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
     }
   }
 
+  /**
+   * Delete the business metadata for the application and the programs.
+   */
+  private void deleteAppBusinessMetadata(Id.Application appId) {
+    // Remove metadata for the Application itself.
+    businessMds.removeMetadata(appId);
+
+    // Remove business metadata for the programs of the Application
+    // TODO: Need to remove this we support prefix search of metadata type.
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    for (ProgramType programType : ProgramType.values()) {
+      deleteProgramBusinessMetadataForApp(programType, appId, appSpec);
+    }
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
+   */
+  private void deleteProgramBusinessMetadataForApp(ProgramType programType, Id.Application appId,
+                                                   ApplicationSpecification appSpec) {
+    Set<String> programNames = null;
+    switch (programType) {
```

Comment:
seems a bit odd to loop through program types and then switch on them.  For each app delete call, there will be guaranteed 2 log info messages about webapp and custom action.  Simpler if we get all flows, add to a `Set<Id.Program>`, get all mapreduce, add to set, etc.